qxf2_scheduler/qxf2_scheduler.py

```diff
@@ -38,7 +50,83 @@ def is_weekend(date):
     date = gcal.process_date_string(date)
     day = date.weekday()
     print('~~~~',day)
-    return True if day==5 or day==6 else False 
+    return True if day==5 or day==6 else False
+
+
+def get_free_slots_in_chunks(free_slots):
+    "Return the free slots in 30 minutes interval"
+    #Appending the 30 minutes slot into list
+    if free_slots == None:
+        print("There are no more free slots available for this user")    
+    else:        
+        chunk_time_interval = []        
+        for i in  range(0,len(free_slots)):
+            #Initializing the free slot end            
+            free_slot_start = free_slots[i]['start']
+            #Intializing the next free slot start        
+            free_slot_end = free_slots[i]['end']
+            
+            #Find the difference between start and end slot
+            diff_between_slots = convert_string_into_time(free_slot_end) - convert_string_into_time(free_slot_start)
+            
+            if diff_between_slots >= timedelta(minutes=30):                
+                if free_slot_start[-2:]=='00':
+                    modified_free_slot_start = free_slot_start                    
+
+                elif free_slot_start[-2:] <= '30':
+                    modified_free_slot_start = '{}:{}'.format(free_slot_start.split(':')[0], '30')
+                    
+                elif free_slot_start[-2:] > '30':
+                    free_slot_start = '{}:{}'.format(free_slot_start.split(':')[0], '00')
+                    modified_free_slot_start = convert_string_into_time(free_slot_start) + timedelta(hours=1)
+                    modified_free_slot_start = get_datetime_in_time_format(modified_free_slot_start)
+                    
+                chunk_slots = modified_free_slot_start
+
+                if free_slot_end[-2:]=='00' or free_slot_end[-2:]=='30' :
+                    modified_free_slot_end = free_slot_end                    
+
+                elif free_slot_end[-2:] <'30':                    
+                    modified_free_slot_end = '{}:{}'.format(free_slot_end.split(':')[0], '00')
+                    
+                elif free_slot_end[-2:] > '30':
+                    modified_free_slot_end = '{}:{}'.format(free_slot_end.split(':')[0], '30')                    
+                
+                result_flag = True
```

Comment:
Lines 95-128 should become a method of its own. Something like `get_chunks_in_slot(slot_start,slot_end, chunk_size):`

That way your `get_free_slots_in_chunks(free_slots)` method looks something like this:

```
def get_free_slots_in_chunks(free_slots,chunk_size=30):
  "Return chunked free slots"
  for free_slot in free_slots:
    chunk_time_intervals = []
    free_slot_start = get_modified_free_slot_start(free_slot['start'],marker=chunk_size)
    free_slot_end = get_modified_free_slot_end(free_slot['end'],marker=chunk_size)
    slot_duration = convert_string_into_time(free_slot_end) - convert_string_into_time(free_slot_start)
    if slot_duration >= chunk_size:
      chunks = get_chunks_in_slot(free_slot_start,free_slot_end, chunk_size=chunk_size)
      chunk_time_intervals += chunks
  
   return chunk_time_intervals
```

Reply:
done